tools/wireshark/plugin.c

```diff
@@ -0,0 +1,894 @@
+		field_info->hfinfo.hfinfo.type = FT_DOUBLE;
+		field_info->hfinfo.hfinfo.display = BASE_NONE;
+	} else if (strcmp(type_str, "boolean") == 0) {
+		ws_info("proto registering key %s as boolean", path);
+		field_info->hfinfo.hfinfo.type = FT_BOOLEAN;
+		field_info->hfinfo.hfinfo.display = BASE_NONE;
+	} else if (strcmp(type_str, "array") == 0) {
+		ws_info("proto registering key %s as array", path);
+		field_info->hfinfo.hfinfo.type = FT_STRINGZ;
+		field_info->hfinfo.hfinfo.display = BASE_NONE;
+	} else if (!strncmp(type_str, "object", 6)) {
+		int *p_ett_id, err;
+
+		ws_info("proto registering key %s as object", path);
+		field_info->hfinfo.hfinfo.type = FT_NONE;
+
+		/* Create and store the pointer to where Wireshark will put the ETT ID */
+		p_ett_id = g_malloc(sizeof(int));
+		*p_ett_id = -1;
+
+		g_hash_table_insert(retis_ett, g_strdup(path), p_ett_id);
+		err = object_parse(def, path, errstr);
+		if (err)
+			return err;
+	} else {
+		g_free(field_info);
+		*errstr = g_strdup_printf("unsupported type: %s: %s", type_str,
+					  json_object_to_json_string(def));
+		return -1;
+	}
+
+	g_hash_table_insert(retis_field_info, g_strdup(path), field_info);
+	return 0;
+}
+
+/* Best-effort attept to generate a "name" for a type. */
+char *definition_get_name(json_object *definition, const char *key)
+{
+	json_object *desc;
+	char *name;
+
+	if (json_object_object_get_ex(definition, "description", &desc) &&
+	    json_object_get_string_len(desc)) {
+		/* Get the first line without full stops. */
+		const char *desc_str = json_object_get_string(desc);
+		size_t off;
+
+		for (off = 0; off < strlen(desc_str); off++) {
+			if (desc_str[off] == '\n' || desc_str[off] == '\r' ||
+			    desc_str[off] == '.')
+				break;
+		}
+		name = g_malloc(off + 1);
+		name[off] = '\0';
+		memcpy(name, desc_str, off);
+	} else {
+		name = g_strdup(key);
+	}
+	return name;
+}
+
+static int object_properties_parse(json_object *props,
+				   const gchar *current_path, gchar **errstr)
+{
+	if (!json_object_is_type(props, json_type_object)) {
+		*errstr = g_strdup_printf("properties is not an object: %s",
+					  json_object_to_json_string(props));
+		return -1;
+	}
+
+	json_object_object_foreach(props, prop_key, prop_value)
+	{
+		gchar *prop_json_path = g_strdup_printf(
+			"%s%s%s", current_path,
+			current_path[0] == '\0' ? "" : ".", prop_key);
+		json_object *any_of, *definition = NULL;
+		int err;
+
+		if (json_object_object_get_ex(prop_value, "anyOf", &any_of)) {
+			for (size_t i = 0; i < json_object_array_length(any_of);
+			     i++) {
+				json_object *ref = resolve_json_reference(
+					json_object_array_get_idx(any_of, i));
+				if (ref) {
+					// Do we need to cache parsed defs??
+					definition = ref;
+					break;
+				}
+			}
+		} else {
+			definition = prop_value;
+		}
+
+		if (!definition) {
+			*errstr = g_strdup_printf(
+				"definition not found on property %s",
+				prop_key);
+			return -1;
```

Comment:
+ `g_free(prop_json_path);`

I'm used to handle those kind of repetitive error paths using goto. It's usually making things a bit more robust over time, this and other functions could be converted. It's probably a matter of taste so feel free to ignore.

Reply:
I don't really care, I can go with goto as well.